bigtable/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/commands/Serve.java

```diff
@@ -0,0 +1,124 @@
+import io.grpc.ServerCallHandler;
+import io.grpc.netty.shaded.io.grpc.netty.NettyServerBuilder;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Help.Visibility;
+import picocli.CommandLine.Option;
+
+@Command(name = "serve", mixinStandardHelpOptions = true, description = "Start the proxy server")
+public class Serve implements Callable<Void> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(Serve.class);
+
+  @Option(
+      names = "--listen-port",
+      required = true,
+      description = "Local port to accept connections on")
+  int listenPort;
+
+  @Option(names = "--useragent", showDefaultValue = Visibility.ALWAYS)
+  String userAgent = "bigtable-java-proxy";
+
+  @Option(
+      names = "--bigtable-data-endpoint",
+      converter = Endpoint.ArgConverter.class,
+      showDefaultValue = Visibility.ALWAYS)
+  Endpoint dataEndpoint = Endpoint.create("bigtable.googleapis.com", 443);
+
+  @Option(
+      names = "--bigtable-admin-endpoint",
+      converter = Endpoint.ArgConverter.class,
+      showDefaultValue = Visibility.ALWAYS)
+  Endpoint adminEndpoint = Endpoint.create("bigtableadmin.googleapis.com", 443);
+
+  ManagedChannel adminChannel = null;
+  ManagedChannel dataChannel = null;
+  Server server;
+
+  @Override
+  public Void call() throws Exception {
+    start();
+    server.awaitTermination();
+    cleanup();
+    return null;
+  }
+
+  void start() throws IOException {
+    if (dataChannel == null) {
+      dataChannel =
+          ManagedChannelBuilder.forAddress(dataEndpoint.getName(), dataEndpoint.getPort())
+              .maxInboundMessageSize(256 * 1024 * 1024)
+              .disableRetry()
+              .keepAliveTime(30, TimeUnit.SECONDS)
+              .keepAliveTimeout(10, TimeUnit.SECONDS)
+              .build();
+    }
+    if (adminChannel == null) {
+      adminChannel =
+          ManagedChannelBuilder.forAddress(adminEndpoint.getName(), adminEndpoint.getPort())
+              .userAgent(userAgent)
+              .disableRetry()
+              .build();
+    }
+
+    Map<String, ServerCallHandler<byte[], byte[]>> serviceMap =
+        ImmutableMap.of(
+            BigtableGrpc.SERVICE_NAME,
+            new ProxyHandler<>(dataChannel),
+            BigtableInstanceAdminGrpc.SERVICE_NAME,
+            new ProxyHandler<>(adminChannel),
+            BigtableTableAdminGrpc.SERVICE_NAME,
+            new ProxyHandler<>(adminChannel),
+            OperationsGrpc.SERVICE_NAME,
+            new ProxyHandler<>(adminChannel));
+
+    server =
+        NettyServerBuilder.forAddress(
+                new InetSocketAddress("localhost", listenPort), InsecureServerCredentials.create())
+            .fallbackHandlerRegistry(new Registry(serviceMap))
```

Comment:
Why do we use fallback registry instead of adding service directly?

Reply:
Because I want the implementation to be generic. If I used concrete services, then I would have implement every method like:

```
class MyBigtableInstanceAdmin extends BigtableInstanceAdminImplBase {
   void createInstance(CreateInstanceRequest req, ResponseObserver<Operation> observer) {
      proxyToUpstream(BigtableInstanceAdminGrpc.getCreateInstanceMethodDescriptor(), req, observer);
   }
   void getInstance(GetInstanceRequest req, ResponseObserver<Instance> observer) {
      proxyToUpstream(BigtableInstanceAdminGrpc.getGetMethodDescriptor(), req, observer);
   }
  ....
}
```